packages/tpetra/core/src/Tpetra_MultiVector_def.hpp

```diff
@@ -4357,66 +4357,46 @@ namespace Tpetra {
          "this bug to the Tpetra developers.");
     }
 
-    // MPI requires that the send and receive buffers don't alias one
-    // another, so we have to copy temporary storage for the result.
-    //
-    // We expect that MPI implementations will know how to read device
-    // pointers.
-    dev_view_type tgtBuf_dev;
-    host_view_type tgtBuf_host;
-    if (useHostVersion) {
-      tgtBuf_host = host_view_type ("tgtBuf", lclNumRows, numCols);
-    }
-    else {
-      tgtBuf_dev = dev_view_type ("tgtBuf", lclNumRows, numCols);
-    }
-
     const int reduceCount = static_cast<int> (totalAllocSize);
     if (useHostVersion) {
-      TEUCHOS_TEST_FOR_EXCEPTION
-        (static_cast<size_t> (tgtBuf_host.size ()) < totalAllocSize,
-         std::logic_error, "Tpetra::MultiVector::reduce: tgtBuf_host.size() = "
-         << tgtBuf_host.size () << " < lclNumRows*numCols = " << totalAllocSize
-         << ".  Please report this bug to the Tpetra developers.");
       reduceAll<int, impl_scalar_type> (comm, REDUCE_SUM, reduceCount,
                                         srcBuf_host.data (),
-                                        tgtBuf_host.data ());
+                                        srcBuf_host.data ());
```

Comment:
This is only correct if the MPI communicator is _not_ an intercommunicator.  Tpetra has a predicate for that: https://github.com/trilinos/Trilinos/blob/69f8850c25bdead51b38866ae34d4fa40947b67f/packages/tpetra/core/src/Tpetra_Details_isInterComm.hpp#L82

```c++
if (Details::isInterComm (comm)) {
  // intercomms are rare; don't prematurely optimize
  // 
  // allocate temp tgtBuf_host 
  // reduceAll from srcBuf_host to tgtBuf_host
  // copy back to srcBuf_host
}
else {
  reduceAll<int, impl_scalar_type> (comm, REDUCE_SUM, reduceCount, srcBuf_host.data (), srcBuf_host.data ());
}
```

Reply:
That's probably safer, but shouldn't we always be in that case? This is used in `MV::multiply`, and uses the communicator of the result C.

Reply:
@cgcgcg It's always legal to create a Map with an Comm that wraps an intercommunicator.  Intercommunicators behave like intracommunicators (the usual ones, including `MPI_COMM_WORLD`), except for this "`MPI_IN_PLACE` doesn't work" property.

Reply:
@mhoemmen Sounds good. I'll change it.